### ISSUE TITLE
React Native on macOS or on web - broken width and month not changing

### ISSUE THREAD

**Issue by rdewolff** [hoangnm/react-native-week-view#198]
The library is not working when compiling a react-native project for macOS (Catalyst for example).
It's not working either when using react-native-web version.

The following elements are not working:
1. The width of the day header is not correct
2. The scroll is hard to test in these conditions, but it seems to not be able to trigger a next month change when you scroll past a certain amount of time horizontally.

Please find a video illustrating the first examples:

![react-native-web-week-view-480](https://user-images.githubusercontent.com/130704/166927444-09dbcdde-7a55-4847-b7b8-a2d8fe9054de.gif)

Steps to reproduce:
- create a react-native-web repository (u can use boilerplates like https://github.com/infinitered/ignite that support iOS/android **and web**
- add `react-native-week-view` package
- fix the issue reported on https://github.com/hoangnm/react-native-week-view/issues/196
- include the `<WeekView... />` component on a page
- try to render it on a web page

**Comment by pdpino** [comment]
@rdewolff I'll check this out ASAP

In the broken web app, where are events placed in the screen? For example an event in the central week, the week before and the week after.
Could you show a screenshot/gif with those?

**Comment by rdewolff** [comment]
I haven't added meets on the first test I did. Let me give it a try and get back to u

**Comment by pdpino** [comment]
@rdewolff I was able to reproduce it and display some dummy events, so don't worry:

![image](https://user-images.githubusercontent.com/20805451/167113511-5b901cf3-7110-495f-bbd8-82ead69f425f.png)

**Comment by rdewolff** [comment]
Aaah great, you got a setup working 🥳

**Comment by rdewolff** [comment]
I just tried to display some events in the grid but did not get it working. Did you already do some fixes @pdpino to get your event displayed?

**Comment by pdpino** [comment]
**TL;DR**: I fixed some things, but is still not working 100%, _probably_ due to an upstream bug from react-native-web `VirtualizedList`. Demo:

![fixed-header](https://user-images.githubusercontent.com/20805451/167203006-455f6d94-af67-4027-8989-1c10cd7646ae.gif)

(btw the cursor is not moving because I was scrolling with the wheel)

---

**Details:**

There are multiple things in the initial example that went wrong, I was able to fix some of them:

1. Header does not have the appropriate width
2. All the grey grid lines are rendered at 00:00hrs
3. Vertical scrolling is not working properly, since in web it may require some tricks
  * You can scroll, but the initial scroll to the `startHour` is not done (i.e. you cannot call `scrollTo()` for the `ScrollView` ref)
4. Horizontal scrolling is also broken
  4.1 The initial position should be today (in my example would be May 6), but the initial position shows April 21
  4.2 You cannot scroll past certain dates to the right or left
5. When scrolling the grid horizontally, the header does not follow the grid smoothly (i.e. animated)

What I did:

1. :+1: Compute the header width using the size of the window
    * required working on top of this recent PR https://github.com/hoangnm/react-native-week-view/pull/189
2. :+1: Grey gridlines are now properly rendered throughout the grid (was a simple flex issue)
3. :+1: The initial vertical position is fixed, and the vertical scroll works properly
    * Note: this required constraining the `ScrollView`'s height to `height: '100vh'` (100 viewport's height, web only)
4. `VirtualizedList` and `FlatList` have shown issues in react-native-web (for example see a list of issues here: https://github.com/necolas/react-native-web/pull/2241#issuecomment-1055824946)
    * 4.1. :+1: The `initialScrollIndex` behavior could be fixed --> the week starts pointing to today (see gif)
      * I needed to install the 0.18 preview version of react-native-web (`yarn add react-native-web@0.0.0-b2a3e86d`, see preview here: https://github.com/necolas/react-native-web/pull/2248)  
    *  4.2 :-1: :-1: The scrolling is still constrained to 5 pages (weeks), from April 18 to May 22 see gif
        * Culprit: the `onMomentumScrollEnd` event from the horizontal `VirtualizedList` is not being triggered
        * When the user scrolls, we use that event to add pages (weeks) to the left and right
        * At the start, we render 5 pages (those are the only pages that we can see)
        * I tested a small `VirtualizedList` example, and I think this is an issue from react-native-web (I can provide more details for a small repro)
5. :-1: Could not make the animation smoother (though, the horizontal scrolling should be resolved first anyway)

4.1 is probably a dealbreaker for most apps. We should corroborate if it is a bug from react-native-web; hopefully they will have a workaround soon

**Comment by pdpino** [comment]
> Did you already do some fixes to get your event displayed?

@rdewolff tbh I don't remember now if I had already made a small fix :laughing:. I made a draft PR with the changes so far

**Comment by rdewolff** [comment]
Concerning 4.2, is there a way to use an alternative to `onMomentumScrollEnd`?

**Comment by pdpino** [comment]
There are no good options that I can think of

* `onScroll` event
  * Is triggered too often (once per frame, used for UI events) --> could lag the whole swiping interaction
  * Currently we use it to sync the grid with the header (bullet 5 mentioned)
* `onEndReached` event
  * This could work on one side of the calendar (left or right)
  * but there is no `onStartReached` --> would not work on the other side (right or left)
* I've seen libraries that implement bidirectional infinite scrolling, but they do not support horizontal: https://github.com/GetStream/react-native-bidirectional-infinite-scroll/issues/5

**Comment by rdewolff** [comment]
So our best bet might be to wait for `react-native-web` to support `onMomentumScrollEnd`.

Maybe this comment is related to our issue: https://github.com/expo/expo/issues/16822#issuecomment-1099169366 ?

**Comment by rdewolff** [comment]
How can we get this issue moving forward?

**Comment by pdpino** [comment]
> How can we get this issue moving forward?

PR #201 will fix up to 4.1.

As to 4.2, I'm afraid without scroll events in web there is not much more we can do. I see a lot of attention in rn-web, (e.g. https://github.com/necolas/react-native-web/issues/2249, https://github.com/necolas/react-native-web/pull/2241), hopefully they will have a fix soon

On the side, I'm implementing some performance improvements, and that _might_ include migrating to recycler-list-view, which _should_ support rn-web scroll events. Although these changes are taking me a bit longer. (I'll probably open a separate issue regarding this soon)

**Comment by rdewolff** [comment]
PR #201 was merged 🥳

**Comment by pdpino** [comment]
@rdewolff necolas (rn-web 's author) [commented:](https://github.com/necolas/react-native-web/issues/2249#issuecomment-1150434795)

> those of you who need these events will need to implement and submit PRs for. I do not plan to work on adding these features myself.

I cannot work on that right now.

---

To recap, I think the options are:
1. wait for react-native-web to support scroll events

or wait for other libraries, e.g.:

2. [react-native-bidirectional-infinite-scroll](https://github.com/GetStream/react-native-bidirectional-infinite-scroll),  does not [support horizontal](https://github.com/GetStream/react-native-bidirectional-infinite-scroll/issues/5)
3. [recyclerlistview](https://github.com/Flipkart/recyclerlistview), needs [bidirectional scrolling](https://github.com/Flipkart/recyclerlistview/pull/647)

with 2 or 3 we could try to make the change only for web (instead of web and mobile), if that's simpler. In any case, have in mind is a big refactor and may cause conflicts with other stuff (animations, etc)

**Comment by rdewolff** [comment]
The current workaround is to add navigation button to navigate to previous or next week.

**Comment by rdewolff** [comment]
Just found this https://github.com/Flipkart/recyclerlistview
Is it something we could use @pdpino ?

**Comment by pdpino** [comment]
Sorry for the late reply.

@rdewolff recyclerlistview does not support bidirectional infinite scrolling: https://github.com/Flipkart/recyclerlistview/pull/647#issuecomment-1258695663

They implemented a fix for VirtualizedList, and is published in a fork of react-native: https://github.com/facebook/react-native/pull/29466#issuecomment-1255383709. Using this _might_ fix the problem?
